edited on your comments Security.md and Intro.md (#297)
subspace/subspace-docs: versioned_docs/version-latest/protocol/security.md

---
title: General and server security guidelines

> sidebar_position: 5

Reviewer: We should keep the sidebar position at 6 since now we've added the information for timekeepers that should come earlier.

Author: ok i see. So for my two pages -Security and Partitioning - I can count on slots 6 and 7 ?

Reviewer: Let's do 6 and 8 please, 7 could be "Common problems"

description: General security measures. Cyberattacks and fraud protection. Server protection.
keywords:
  - Security
  - Wallet
  - Password
  - Server
  - Protection
  - Fraud
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import Link from '@docusaurus/Link';
import styles from '@site/src/pages/index.module.css';

> Navigating the internet and being part of a crypto project can be exciting! But it's crucial to use best security practices. Here is a general guide to security and safety advice, including basic measures for securing servers and home PCs. This is your contribution to keeping Subspace network up and running and of course the safety of the community's cryptoassets. 

Reviewer: ```suggestion
Navigating the internet and being part of a crypto project can be exciting! But it's crucial to use the best security practices. Here is a general guide to security and safety, including basic measures for securing servers and home PCs. This is your contribution to keeping Subspace network up and running and of course the safety of the community's cryptoassets. 
```

Enjoy Subspace journey with peace of mind!

> # Wallet security

Reviewer: ```suggestion
## Wallet security
```

Author: yes. structure !

### Passwords
*

>  Be sure to make the password long enough, combining both registers letters, numbers and special characters. An 11-character password increases the bruteforce attack time to normally impossible limit but please take a look at [passwords brutforce time illustrating table](https://www.techrepublic.com/wp-content/uploads/2022/03/tr8423-Figure-A-How-an-8-character-password-could-be-cracked-in-just-a-few-minutes-770x404.jpg). However, even a complex password is not all, that is required for proper protection. This will be discussed in this article.

Reviewer: ```suggestion
 Be sure to make the password long enough, combining both registers letters, numbers, and special characters. An 11-character password increases the bruteforce attack time to normally impossible limit. Take a look at [passwords brutforce time illustrating table](https://www.techrepublic.com/wp-content/uploads/2022/03/tr8423-Figure-A-How-an-8-character-password-could-be-cracked-in-just-a-few-minutes-770x404.jpg). However, even a complex password is not all, that is required for proper protection.
```

*
 Avoid using common or easily guessed combinations of words or letters, as well as parts of your online data or reusing any of your existing passwords.

Weak password example:
```
S0methin9C00l!
```
Strong password example:
```
^p$O_~a!4h{G'9C*
```
*

>  Consider using a password manager to generate complex passwords and store them securely. There are plenty of password managers available. Ensure that your choice stores data in encrypted form, boasts a robust password generator, receives positive reviews, and has a history of consistent updates. **For example, well-regarded options include KeePass, Bitwarden, 1Password, LastPass, Dashlane, and Keeper.**

Reviewer: ```suggestion
 Consider using a password manager to generate complex passwords and store them securely. There are plenty of password managers available. Ensure that the password manager of your choice stores data in encrypted form, boasts a robust password generator, receives positive reviews, and has a history of consistent updates. **For example, well-regarded options include KeePass, Bitwarden, 1Password, LastPass, Dashlane, and Keeper.**
```

 However, it's crucial to do your own research and consult ratings before making a decision.
*
 Do not share your 12-word seed phrase or private keys.
*
 A hardware wallet is by far the most secure way to store your private keys. Unlike online exchanges and wallets, hardware wallets store the keys offline and protect from potential malware or hacking attempts.
The most popular hardware wallets are [Ledger](https://www.ledger.com/) and [Tresor](https://trezor.io/).
*

>  Backup your wallets: store the backups securely offline or in encrypted cloud storage. This ensures hat you can recover your funds in case of device loss, damage, or theft. There are also many encrypted cloud storage services available, like Tresorit, pCloud, Sync.com, SpiderOak, Mega (order does not matter). But always do your own research.

Reviewer: ```suggestion
 Backup your wallets: store the backups securely offline or in encrypted cloud storage. This ensures that you can recover your funds in case of device loss, damage, or theft. There are also many encrypted cloud storage services available, like Tresorit, pCloud, Sync.com, SpiderOak, Mega (order does not matter). But always do your own research.
```

*
 Be cautious of public Wi-Fi.
*
 Keep all software up to date.
*
 If possible, enable 2FA (two-factor authentication) for all your cryptocurrency accounts and wallets. This provides an extra layer of security by requiring a second verification step, such as a code from a mobile app or hardware token, in addition to a password.
*

>  Secure your devices! And here's a little basic guide for servers and home PCs.

Reviewer: I think we can remove that sentence.

> # Community security

Reviewer: ```suggestion
## Community security
```

*
 Remember: Subspace team members will never send you any direct messages first. If this happens to you, ignore the message, block the sender, and report the incident to our ["scam report" channel](https://discord.com/channels/864285291518361610/982023010405990400)
*
 Be ware of fishing attacks.
Do not click on any links on our forum, Discord or Telegram, unless these links are posted by one of the team members (ambassadors or staff). Scammers may create fake links looking like legit websites, so the easiest way to stay safe is to avoid such messages at all.

> # Server securing

Reviewer: ```suggestion
## Server securing
```

Together, SSH and RSA provide secure, authenticated, and encrypted connections to protect sensitive information. For added security, you can create an SSH key with a passphrase.

### Creating RSA key pair
#### On the server, run: 
`vi /etc/ssh/sshd_config`
`--> PubkeyAuthentication yes`

`systemctl restart ssh`

#### Create rsa keys on home PC (if you don't have it yet).
```console
ssh-keygen
```
*you will get two keys:*

                        `$HOME/.ssh/id_rsa`     Your private RSA key to keep on local PC

                        `$HOME/.ssh/id_rsa.pub` Your public RSA key to send to a server

#### Transfer rsa keys to the server:

On a home PC: 

`sudo ssh-copy-id -p 12345 user_name@server-ip-addres`

This utility specifically designed for copying SSH keys to a remote server. 
It automatically handles the key placement and permissions on the remote server, making it more convenient. 
Use `-p` flag to specify not standard port, if you have changed it.

> if you have created keys before, don't overwrite it! As you will not be able to authenticate using the previous key anymore. But you can keep them somewhere else and generate next.

Reviewer: ```suggestion
If you have created keys before, don't overwrite it! As you will not be able to authenticate using the previous key anymore. But you can keep them somewhere else and generate them next.
```

Author: Thank you. I have improved my English with the help of your comments!

Alternative ways to transfer RSA public key to a remote server:
*
If you have created keys before and store them elsewhere, you can use rsync to copy the contents of the public key from any other place to `authorized_keys` by specifying the path to the keys:
```console
sudo rsync -e "ssh -p 12345" ~/.ssh/user2/id_rsa.pub USER@SERVER_IP:~/.ssh/authorized_keys
```		
Use -p flag for specifying not standatr port
*
This command will create an .ssh dir on a server(or skip if it has) and add the keys to the end of `authorized_keys` file:
```console
cat ~/.ssh/id_rsa.pub | ssh username@remote_host "mkdir -p ~/.ssh && cat >> ~/.ssh/authorized_keys"
``````
-p flag make the tool to skip the error if the directory already exists
*
Manually transfer the public key file and add it to the `authorized_keys` file on the remote server. 
You can open `id_rsa.pub key` and copy-paste it's content to the end of `authorized_keys` file:
```console
echo public_key_string >> ~/.ssh/authorized_keys
```
:::note 
Make sure you have permissions 700 for `~/.ssh` directory and 600 for `authorized_keys` in it
:::

> #### Test secure connection to a server

Reviewer: ```suggestion
### Test secure connection to a server
```

```console
ssh username@your-server-ip-addr
```

### Managing connections can be much simplier and more enjoyable by creating aliases!
Create a file named "config" in the `~/.ssh` directory (where the keys are), simply add your server or other users for the same server to it, like in example below. Feel free to add as many as you want.

```console
Host subspace                            # Just write here the any name of the alias (for ex. "subspace" or "farmer" etc..)
    HostName 123.123.123.123     # Example. Put you server's IP
    Port 12345                                 # your custom port
    User username			            # User for login (in a rescue system here must be root, change it later)
```

> *indentation aren't important, use 4 spaces for readability*

Reviewer: ```suggestion
*indentation isn't important, 4 spaces indentation is used for a better readability*
```
